### PR TITLE
Run HaplotypeCallerSpark on exome BAM

### PR DESCRIPTION
**Pull request by tomwhite**
These are the changes I needed to get HaplotypeCallerSpark working on an exome BAM.

@lbergelson, could you take a look?

### CODE REVIEW

**Comment by tomwhite** [comment]
Updated to fix the tests. GVCF is not yet supported by VariantsSparkSink in GATK (https://github.com/broadinstitute/gatk/issues/2738), so I changed the write path back to using the inefficient (but correct) existing code.

**Comment by tomwhite** [comment]
Tests now passing. @lbergelson can you take a look?

File: src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java [hunk 1]
```diff
@@ -222,13 +239,19 @@ private void writeVariants(JavaRDD<VariantContext> variants) {
...
      */
-    private static JavaRDD<Shard<GATKRead>> createReadShards(final Broadcast<OverlapDetector<ShardBoundary>> shardBoundariesBroadcast, final JavaRDD<GATKRead> reads) {
+    private static JavaRDD<Shard<GATKRead>> createReadShards(final Broadcast<OverlapDetector<ShardBoundary>> shardBoundariesBroadcast, final JavaRDD<GATKRead> reads, final SAMFileHeader header) {
         final JavaPairRDD<ShardBoundary, GATKRead> paired = reads.flatMapToPair(read -> {
             final Collection<ShardBoundary> overlappingShards = shardBoundariesBroadcast.value().getOverlaps(read);
             return overlappingShards.stream().map(key -> new Tuple2<>(key, read)).iterator();
         });
         final JavaPairRDD<ShardBoundary, Iterable<GATKRead>> shardsWithReads = paired.groupByKey();
-        return shardsWithReads.map(shard -> new SparkReadShard(shard._1(), shard._2()));
+        return shardsWithReads.map(shard -> new SparkReadShard(shard._1(), sortReadsInMemory(shard._2(), header)));
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java]: @tomwhite Why do we need to sort the reads here?  Shouldn't the reads be sorted within shards if they input file was sorted?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java [hunk 2]
```diff
@@ -361,7 +360,17 @@ public void writeHeader( final VariantContextWriter vcfWriter, final SAMSequence
...
         final VCFHeader vcfHeader = new VCFHeader(headerInfo, sampleSet);
         vcfHeader.setSequenceDictionary(sequenceDictionary);
-        vcfWriter.writeHeader(vcfHeader);
+        return vcfHeader;
+    }
+
+    /**
+     * Writes an appropriate VCF header, given our arguments, to the provided writer
+     *
+     * @param vcfWriter writer to which the header should be written
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/HaplotypeCallerEngine.java]: Add javadoc for `sequenceDictionary`

**Review by lbergelson — COMMENTED** [review]
This looks good to me, except for the part about sorting.  I don't understand why we have to do that.  I thought hadoop bam made a guarantee that we got the shards in sorted order?  I assume you ran into some out of order reads, but maybe that's a bug somewhere else?

**Comment by droazen** [comment]
@tomwhite Migration instructions for this branch: https://github.com/broadinstitute/gatk/wiki/Migrating-branches-from-gatk-protected-to-gatk

**Comment by tomwhite** [comment]
Fixed in https://github.com/broadinstitute/gatk/pull/3105.

**Pull request closed without merging**